numpy/_core/src/multiarray/arrayobject.c

```diff
@@ -418,8 +418,9 @@ array_dealloc(PyArrayObject *self)
          */
         Py_XDECREF(fa->base);
     }
-
-    if ((fa->flags & NPY_ARRAY_OWNDATA) && fa->data) {
+    /* Deallocate data, if present and not a small allocation in the object */
+    if ((fa->flags & NPY_ARRAY_OWNDATA) && fa->data != NULL
+        && fa->data != (void *)((char *)fa + PyArray_Type.tp_basicsize)) {
```

Comment:
We could add a (private) flag as well, although maybe this is check is nice enough.

Reply:
Yes, I realized the same. How about setting `mem_handler = Py_None`? (cannot use `NULL` as that is interpreted as meaning that `data` might have been gotten with `malloc`... see around the changes to `array_dealloc`).

Reply:
Hmm, `mem_handler = Py_None` is probably asking for problems, as surely there is code that just acts on it being non-NULL. One would need a no-op function or so. But that becomes more complicated than just having a new flag...

Reply:
So, right now I went for a semi-private `_PyArray_DATA_ON_INSTANCE` in-line function in `ndarraytypes.h` - that way we can always decide later whether or not to define a flag, etc.